File: solvers/java/src/test/java/com/analog/lyric/dimple/test/core/parameterizedMessages/TestParameterizedMessage.java
/*******************************************************************************
*   Copyright 2014 Analog Devices, Inc.
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
********************************************************************************/

package com.analog.lyric.dimple.test.core.parameterizedMessages;

import static com.analog.lyric.util.test.ExceptionTester.*;
import static org.junit.Assert.*;

import java.io.PrintStream;

import org.eclipse.jdt.annotation.Nullable;

import com.analog.lyric.dimple.data.DataRepresentationType;
import com.analog.lyric.dimple.model.values.Value;
import com.analog.lyric.dimple.solvers.core.parameterizedMessages.IParameterizedMessage;
import com.analog.lyric.dimple.solvers.core.parameterizedMessages.ParameterizedMessageBase;
import com.analog.lyric.dimple.test.DimpleTestBase;
import com.analog.lyric.util.test.SerializationTester;
import com.google.common.primitives.Doubles;

/**
 * Base class for IParameterizedMessage tests
 * 
 * @since 0.06
 * @author Christopher Barber
 */
public class TestParameterizedMessage extends DimpleTestBase
{
	private static class BogusParameters extends ParameterizedMessageBase
	{
		private static final long serialVersionUID = 1L;

		@Override
		public BogusParameters clone()
		{
			return new BogusParameters();
		}
		
		@Override
		public boolean objectEquals(@Nullable Object other)
		{
			return other instanceof BogusParameters;
		}

		@Override
		public void print(PrintStream out, int verbosity)
		{
		}

		@Override
		public double evalEnergy(Value value)
		{
			return 0;
		}
		
		@Override
		public double computeKLDivergence(IParameterizedMessage that)
		{
			return 0;
		}

		@Override
		public boolean isNull()
		{
			return true;
		}
		
		@Override
		public void setFrom(IParameterizedMessage other)
		{
		}

		@Override
		public void setUniform()
		{
		}

		@Override
		protected double computeNormalizationEnergy()
		{
			return 0;
		}
	}
	
	public void assertGenericInvariants(IParameterizedMessage message)
	{
		assertTrue(message.objectEquals(message));
		assertFalse(message.objectEquals("bogus"));
		assertEquals(DataRepresentationType.MESSAGE, message.representationType());
		
		IParameterizedMessage clone1 = message.clone();
		assertNotSame(clone1, message);
		assertTrue(message.objectEquals(clone1));
		double divergence = message.computeKLDivergence(clone1);
		assertEquals(0.0, divergence, 1e-9);
		assertTrue(divergence >= 0.0);

		IParameterizedMessage clone2 = SerializationTester.clone(message);
		assertNotSame(clone2, message);
		assertTrue(message.objectEquals(clone2));
		divergence = message.computeKLDivergence(clone2);
		assertEquals(0.0, divergence, 1e-9);
		assertTrue(divergence >= 0.0);
		
		double prevEnergy = clone2.getNormalizationEnergy();
		if (Doubles.isFinite(prevEnergy))
		{
			clone2.addNormalizationEnergy(1.0);
			assertEquals(1.0 + prevEnergy, clone2.getNormalizationEnergy(), 1e-15);
			assertFalse(message.objectEquals(clone2));
		}

		clone2.setNull();
		assertTrue(clone2.isNull());
		
		if (!message.isNull())
		{
			assertFalse(message.objectEquals(clone2));
		}
		
		expectThrow(IllegalArgumentException.class, message, "computeKLDivergence", new BogusParameters());
		
		assertEquals("", message.toString(-1));
		for (int verbosity = 0, prevLength = 0; verbosity < 3; ++verbosity)
		{
			String desc = message.toString(verbosity);
			assertTrue(desc.length() >= prevLength);
			prevLength = desc.length();
		}
		
		clone2.setFrom(message);
		assertTrue(message.objectEquals(clone2));
	}
}

File: modelers/matlab/tests/core/testParameterizedMessage.m
% testParameterizedMessage unit tests for ParameterizedMessage classes
%
% See also ParameterizedMessage, DiscreteMessage

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   Copyright 2012 Analog Devices, Inc.
%
%   Licensed under the Apache License, Version 2.0 (the "License");
%   you may not use this file except in compliance with the License.
%   You may obtain a copy of the License at
%
%       http://www.apache.org/licenses/LICENSE-2.0
%
%   Unless required by applicable law or agreed to in writing, software
%   distributed under the License is distributed on an "AS IS" BASIS,
%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%   See the License for the specific language governing permissions and
%   limitations under the License.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function testParameterizedMessage()
    testDiscreteMessage();
end

function testDiscreteMessage()
    dm1 = DiscreteMessage([.4 .6]);
    assertDiscreteMessageInvariants(dm1);
    assertEqual([.4; .6], dm1.Weight);
    
    dm2 = DiscreteMessage(4);
    assertDiscreteMessageInvariants(dm2);
    assertEqual(ones(4,1), dm2(:));
    
    dm3 = DiscreteMessage([1 2], 'energy');
    assertDiscreteMessageInvariants(dm3);
    assertEqual([1 2]', dm3.Energy);
    
    dm4 = DiscreteMessage([2 3 4], 'ener');
    assertDiscreteMessageInvariants(dm4);
    assertEqual([2 3 4]', dm4.Energy);

    assertExceptionThrown(@() DiscreteMessage(), 'MATLAB:narginchk:notEnoughInputs');
    assertExceptionThrown(@() DiscreteMessage([1 2], 'wieght'), 'MATLAB:unrecognizedStringChoice');
    assertExceptionThrown(@() DiscreteMessage([1 NaN]), 'MATLAB:expectedNonNaN');
    assertExceptionThrown(@() DiscreteMessage([1 2; 3 4]), 'MATLAB:expectedVector');
    assertExceptionThrown(@() DiscreteMessage([]), 'MATLAB:expectedVector');
    assertExceptionThrown(@() DiscreteMessage(['foo', 'bar']), 'MATLAB:invalidType');
    assertExceptionThrown(@() DiscreteMessage(1), '', 'Size must be greater than one.');
end

function assertParameterizedMessageInvariants(message)
    assertTrue(isa(message, 'ParameterizedMessage'));
    
    assertTrue(isa(message.IParameters, ...
        'com.analog.lyric.dimple.solvers.core.parameterizedMessages.IParameterizedMessage'));
    assertEqual(message.IParameters, message.getProxyObject());
    assertEqual(message.IParameters, unwrapProxyObject(message));
end

function assertDiscreteMessageInvariants(message)
    assertParameterizedMessageInvariants(message);
    
    assertTrue(isa(message.IParameters, ...
        'com.analog.lyric.dimple.solvers.core.parameterizedMessages.DiscreteMessage'));
    
    energy = message.Energy;
    weight = message.Weight;
    
    assertEqual(weight, message(:));
    assertEqual(weight(1), message(1));
    assertEqual(weight(1:2), message(1:2));
    assertElementsAlmostEqual(energy, -log(weight));
    assertEqual(numel(weight), message.end(1,2));
    
    message2 = DiscreteMessage(weight);
    assertEqual(weight, message2.Weight);
    message2(1) = 42;
    assertEqual(42, message2(1));
    assertEqual(42, message2.Weight(1));
    assertEqual(weight(2:end), message2(2:end));
    
    message3 = DiscreteMessage(message.IParameters);
    assertEqual(message.IParameters, message3.IParameters);
end

